api/src/util/text_extractor.py

```diff
@@ -0,0 +1,157 @@
+    text_runs = []
+    for slide in presentation.slides:
+        for shape in slide.shapes:
+            if not shape.has_text_frame:
+                continue
+            for paragraph in shape.text_frame.paragraphs:
+                for run in paragraph.runs:
+                    if text := run.text.strip():
+                        text_runs.append(text)
+    data = "\n".join(text_runs)
+    stream.close()
+    return data
+
+
+def extract_text_from_rft(rtf_data: str) -> str:
+    return rtf_to_text(rtf_data)
+
+
+def extract_text_from_docx(file_data: bytes) -> str:
+    return "\n".join(paragraph.text for paragraph in Document(BytesIO(file_data)).paragraphs)
+
+
+def xls_reader(file_path: str) -> str:
+    # Get all rows from every sheet in the xlsx file.
+    all_data = {}
+    excel_file = pandas.ExcelFile(file_path)
+    for sheet_name in excel_file.sheet_names:
+        xls_data = pandas.read_excel(excel_file, sheet_name=sheet_name, dtype=str)
+        all_data[sheet_name] = xls_data.astype(str).values.tolist()
+    return _xls_dict_to_string(all_data)
+
+
+@dataclass
+class TextExtractorConfig:
+    extractor: Callable[[bytes], str] | Callable[[str], str] = lambda data: data
+    read_mode: str = "r"
+    reader: Callable[[str], bytes | str] | None = None
+
+
+class TextExtractor:
+    def __init__(
+        self,
+        file_path: str,
+        file_type: str | None = None,
+    ) -> None:
+        self.file_path = file_path
+        self.file_type = file_type.lower() if file_type else self.file_path.split(".")[-1].lower()
+        self._validate_file_type()
+        self.config = TextExtractor.get_configs()[self.file_type]
+
+    def get_text(self) -> str:
+        try:
+            return self.config.extractor(self._read_file_data()).strip()
+        except Exception as e:
+            err = f"TextExtractorError: Could not extract text from file {self.file_path} - {self.file_type}: {e}"
+            logger.exception(err)
+            raise TextExtractorError(err) from e
+
+    @staticmethod
+    def get_configs() -> dict:
+        html_extractor_config = TextExtractorConfig(extractor=extract_text_from_html)
+        xls_extractor_config = TextExtractorConfig(reader=xls_reader)
+        return {
+            "docx": TextExtractorConfig(extractor=extract_text_from_docx, read_mode="rb"),
+            "pdf": TextExtractorConfig(extractor=extract_text_from_pdf, read_mode="rb"),
+            "pptx": TextExtractorConfig(extractor=extract_text_from_pptx, read_mode="rb"),
+            "txt": TextExtractorConfig(),
+            "html": html_extractor_config,
+            "htm": html_extractor_config,
+            "rtf": TextExtractorConfig(extractor=extract_text_from_rft),
+            "xlsm": xls_extractor_config,
+            "xlsx": xls_extractor_config,
+        }
+
+    def _read_file_data(self) -> bytes | str:
+        if self.config.reader:
+            return self.config.reader(self.file_path)
```

Comment:
If there is a reader, it also needs to be opened with `open_stream` - we can't pass a file path around as the file path is on s3 and open_stream handles that. If the underlying readers need a file path and can't take in a file object, then we can't use them.